### ISSUE TITLE
[java] UnnecessaryCast reported but cast is needed.

### ISSUE THREAD

**Issue by fesse** [pmd/pmd#5740]
**Affects PMD Version:** 
7.13.0

**Rule:**
UnnecessaryCast
https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#unnecessarycast

**Description:**
The rule falsely marks a cast as unnecessary when a method in another class is used.
It works if the cast is applied to the method's actual result but not if it is done at the row's end result.

**Code Sample demonstrating the issue:**

```java
public class CastProblem {

	static double getDouble() {
		return 1.0;
	}

	public static void main(String[] args) {

		// Local method - OK
		int result = (int) getDouble() * 100;
		result += (int) (getDouble() * 100);

		// Inner class - OK
		OtherInnerClass otherInnerClass = new OtherInnerClass();
		result += (int) otherInnerClass.getDouble() * 100;
		result += (int) (otherInnerClass.getDouble() * 100);

		// External class - reports unnecessary cast
		OtherClass otherClass = new OtherClass();
		// OK when casting the first argument
		result += (int) otherClass.getDouble() * 100;

		// Reports Unnecessary cast when casting the whole expression
		// This is not correct. Removing the cast cause a compilation error
		result += (int) (otherClass.getDouble() * 100);

		// Just a bogus line to get rid of unused variable warning
		System.exit(result > 1 ? 1 : 0);
	}

	private static class OtherInnerClass {

		public double getDouble() {
			return 1.0;
		}
	}
}

public class OtherClass {
	public double getDouble() {
		return 1.0;
	}
}

```

**Expected outcome:**
PMD reports a violation at line: result += (int) (otherClass.getDouble() * 100); , but that's wrong. That's a false positive.

**Running PMD through:** *Gradle*

**Comment by oowekyala** [comment]
> ```java
>  // This is not correct. Removing the cast cause a compilation error
> result += (int) (otherClass.getDouble() * 100);
> ```
It does not cause a compilation error, in fact the cast really is unnecessary here. See [Java spec](https://docs.oracle.com/javase/specs/jls/se22/html/jls-15.html#jls-15.26.2). Be that as it may I think this is not a code style we should recommend anyway, ie we should not recommend removing casts at the RHS of a compound assignment unless the user has opted in, or the cast is not losing precision.

The reason why the rule works differently when you are using a method of an inner class or of an external class is likely that your PMD plugin in gradle is not configured properly. If it were, the rule would behave identically in both cases. This might be a case where you ran the PMD plugin without compiling your sources beforehand.

On the other hand I think the rule should be silent even if some types are unresolved because of missing compiled classes. We should probably propagate the unknown type through the binary expression node.

**Comment by fesse** [comment]
I am so sorry for having wasted your time.

In our gradle configuration, we had the following:

```
tasks.withType(Pmd) {
	threads = 4
	group "Verification"
	classpath = null
	maxHeapSize = "2g"
}

```

I am not sure why we hade the line:
classpath = null

Removing this line makes PDM work as expected.
